### PR TITLE
dynamically move pivot even during chain sync

### PR DESCRIPTION
**Pull request by jeongkyun-oh**
## Proposed changes

- This PR is derived from https://github.com/ethereum/go-ethereum/pull/21529

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [ ] New feature or enhancement
- [x] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: datasync/downloader/downloader.go [hunk 1]
```diff
@@ -827,6 +859,20 @@ func (d *Downloader) fetchHeaders(p *peerConnection, from uint64, pivot uint64)
...
+
+		ttl = d.requestTTL()
+		timeout.Reset(ttl)
+
+		d.pivotLock.RLock()
+		pivot := d.pivotHeader.Number.Uint64()
+		d.pivotLock.RUnlock()
+
+		p.logger.Trace("Fetching next pivot header", "number", pivot+uint64(fsMinFullBlocks))
+		go p.peer.RequestHeadersByNumber(pivot+uint64(fsMinFullBlocks), 2, fsMinFullBlocks-9, false) // move +64 when it's 2x64-8 deep
```
Comment (hqjang-pepper) [inline datasync/downloader/downloader.go:874]: I couldn't understand this line. What does 2*64-8 means?

File: datasync/downloader/downloader.go [hunk 2]
```diff
@@ -575,41 +587,60 @@ func (d *Downloader) Terminate() {
 	d.Cancel()
 }
 
-// fetchHeight retrieves the head header of the remote peer to aid in estimating
-// the total time a pending synchronisation would take.
-func (d *Downloader) fetchHeight(p *peerConnection) (*types.Header, error) {
+// fetchHead retrieves the head header and prior pivot block (if available) from
+// a remote peer.
+func (d *Downloader) fetchHead(p *peerConnection) (head *types.Header, pivot *types.Header, err error) {
```
[the remark this answers is not in the document]
Comment (jeongkyun-oh) [reply datasync/downloader/downloader.go:592]: Right, thanks :)

**Review by aidan-kwon — DISMISSED** [review]

File: datasync/downloader/downloader.go [hunk 3]
```diff
@@ -447,10 +450,17 @@ func (d *Downloader) syncWithPeer(p *peerConnection, hash common.Hash, td *big.I
...
 
 	// Look up the sync boundaries: the common ancestor and the target block
-	latest, err := d.fetchHeight(p)
+	latest, pivot, err := d.fetchHead(p)
 	if err != nil {
 		return err
 	}
+	if mode == FastSync && pivot == nil {
+		// If no pivot block was returned, the head is below the min full block
+		// threshold (i.e. new chian). In that case we won't really fast sync
```
Comment (kjhman21) [inline datasync/downloader/downloader.go:459]:
```suggestion
		// threshold (i.e. new chain). In that case we won't really fast sync
```

File: datasync/downloader/downloader.go [hunk 4]
```diff
@@ -487,14 +496,17 @@ func (d *Downloader) syncWithPeer(p *peerConnection, hash common.Hash, td *big.I
...
-		func() error { return d.fetchHeaders(p, origin+1, pivot) }, // Headers are always retrieved
-		func() error { return d.fetchBodies(origin + 1) },          // Bodies are retrieved during normal and fast sync
-		func() error { return d.fetchReceipts(origin + 1) },        // Receipts are retrieved during fast sync
-		func() error { return d.fetchStakingInfos(origin + 1) },    // StakingInfos are retrieved during fast sync
-		func() error { return d.processHeaders(origin+1, pivot, td) },
+		func() error { return d.fetchHeaders(p, origin+1) },     // Headers are always retrieved
+		func() error { return d.fetchBodies(origin + 1) },       // Bodies are retrieved during normal and fast sync
+		func() error { return d.fetchReceipts(origin + 1) },     // Receipts are retrieved during fast sync
+		func() error { return d.fetchStakingInfos(origin + 1) }, // StakingInfos are retrieved during fast sync
+		func() error { return d.processHeaders(origin+1, td) },
```
Comment (kjhman21) [inline datasync/downloader/downloader.go:503]:
```suggestion
		func() error { return d.processHeaders(origin + 1, td) },
```

File: datasync/downloader/downloader.go [hunk 5]
```diff
@@ -487,14 +496,17 @@ func (d *Downloader) syncWithPeer(p *peerConnection, hash common.Hash, td *big.I
 	}
 
 	fetchers := []func() error{
-		func() error { return d.fetchHeaders(p, origin+1, pivot) }, // Headers are always retrieved
-		func() error { return d.fetchBodies(origin + 1) },          // Bodies are retrieved during normal and fast sync
-		func() error { return d.fetchReceipts(origin + 1) },        // Receipts are retrieved during fast sync
-		func() error { return d.fetchStakingInfos(origin + 1) },    // StakingInfos are retrieved during fast sync
-		func() error { return d.processHeaders(origin+1, pivot, td) },
+		func() error { return d.fetchHeaders(p, origin+1) },     // Headers are always retrieved
```
Comment (kjhman21) [inline datasync/downloader/downloader.go:499]:
```suggestion
		func() error { return d.fetchHeaders(p, origin + 1) },     // Headers are always retrieved
```

File: datasync/downloader/downloader.go [hunk 6]
```diff
@@ -827,6 +859,20 @@ func (d *Downloader) fetchHeaders(p *peerConnection, from uint64, pivot uint64)
...
 	}
+	getNextPivot := func() {
+		pivoting = true
+		request = time.Now()
+
+		ttl = d.requestTTL()
+		timeout.Reset(ttl)
+
+		d.pivotLock.RLock()
+		pivot := d.pivotHeader.Number.Uint64()
```
Comment (kjhman21) [inline datasync/downloader/downloader.go:870]: Don't we need to check nil here like https://github.com/klaytn/klaytn/pull/1454/files#diff-99163d68e23ed18d94f9b098d41c9d8c23eeb3e71299102c5c25c9a4ef6245e8R897-R899 ?

File: datasync/downloader/downloader.go [hunk 7]
```diff
@@ -1359,6 +1445,13 @@ func (d *Downloader) processHeaders(origin uint64, pivot uint64, td *big.Int) er
 						}
 					}
 					// If we're importing pure headers, verify based on their recentness
+					var pivot uint64
```
Comment (kjhman21) [inline datasync/downloader/downloader.go:1448]: It seems no use of `pivot` variable.

File: datasync/downloader/downloader.go (hunk 3 above)
Comment (jeongkyun-oh) [reply datasync/downloader/downloader.go]: Fixed :)

File: datasync/downloader/downloader.go (hunk 5 above)
Comment (jeongkyun-oh) [reply datasync/downloader/downloader.go:499]: This is weird, but this is what `gofumpt` does.

File: datasync/downloader/downloader.go (hunk 4 above)
Comment (jeongkyun-oh) [reply datasync/downloader/downloader.go:503]: This is weird, but this is what `gofumpt` does.

File: datasync/downloader/downloader.go (hunk 6 above)
Comment (jeongkyun-oh) [reply datasync/downloader/downloader.go:870]: As we talked in f2f, `getNextPivot` is called after checkint `pivot > 0`.

File: datasync/downloader/downloader.go [hunk 8]
```diff
@@ -799,14 +830,15 @@ func (d *Downloader) findAncestor(p *peerConnection, height uint64) (uint64, err
...
-func (d *Downloader) fetchHeaders(p *peerConnection, from uint64, pivot uint64) error {
+func (d *Downloader) fetchHeaders(p *peerConnection, from uint64) error {
 	p.logger.Debug("Directing header downloads", "origin", from)
 	defer func(start time.Time) {
 		p.logger.Debug("Header download terminated", "elapsed", time.Since(start))
 	}(time.Now())
 
 	// Create a timeout timer, and the associated header fetcher
-	skeleton := true            // Skeleton assembly phase or finishing up
+	skeleton := true // Skeleton assembly phase or finishing up
```
Comment (kjhman21) [inline datasync/downloader/downloader.go:840]: This is unnecessary change. Can we add more whitespace before the comment?

File: datasync/downloader/downloader.go (hunk 6 above)
Comment (kjhman21) [inline datasync/downloader/downloader.go:870]: It could be better to use 'pivotNumber' instead of 'pivot' to increase readability. 'pivot' is used somewhere else for an instance of *types.Header.

File: datasync/downloader/downloader.go (hunk 4 above)
Comment (jeongkyun-oh) [reply datasync/downloader/downloader.go:503]: Yes. It looks like as follows for multi params `(a+b, c)`, for single param `(a + b)`

File: datasync/downloader/downloader.go (hunk 8 above)
Comment (jeongkyun-oh) [reply datasync/downloader/downloader.go:840]: This is also what `gofumpt` does. Maybe we can swap the line and the next one to align the comments.

File: datasync/downloader/downloader.go (hunk 7 above)
Comment (jeongkyun-oh) [reply datasync/downloader/downloader.go:1448]: It seems the pivot is used only the following if statement.
L1456 `chunk[len(chunk)-1].Number.Uint64()+uint64(fsHeaderForceVerify) > pivot`

File: datasync/downloader/downloader.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jeongkyun-oh) [reply datasync/downloader/downloader.go:874]: The line you mentioned is two pairs as follows.
1. request: `go p.peer.RequestHeadersByHash(latest, fetch, fsMinFullBlocks-1, true)`
2. check: `if pivot.Number.Uint64() != head.Number.Uint64()-uint64(fsMinFullBlocks)`

And for retrieving next pivot
1. request: `go p.peer.RequestHeadersByNumber(pivot+uint64(fsMinFullBlocks), 2, fsMinFullBlocks-9, false)`
2. check
  - `have, want := headers[0].Number.Uint64(), pivot+uint64(fsMinFullBlocks); have != want`
  - `have, want := headers[1].Number.Uint64(), pivot+2*uint64(fsMinFullBlocks)-8; have != want`

File: datasync/downloader/downloader.go (hunk 8 above)
Comment (kjhman21) [reply datasync/downloader/downloader.go:840]: Ok. got it.

File: datasync/downloader/downloader.go (hunk 7 above)
Comment (kjhman21) [reply datasync/downloader/downloader.go:1448]: got it. Thanks.

File: datasync/downloader/downloader.go (hunk 5 above)
Comment (kjhman21) [reply datasync/downloader/downloader.go:499]: got it .yes.

File: datasync/downloader/downloader.go (hunk 4 above)
Comment (kjhman21) [reply datasync/downloader/downloader.go:503]: got it.

File: datasync/downloader/downloader.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kjhman21) [reply datasync/downloader/downloader.go:874]: Okay. I thought that the first check condition is used universally, but it is kind of a sequential process.

File: datasync/downloader/downloader.go (hunk 6 above)
Comment (jeongkyun-oh) [reply datasync/downloader/downloader.go]: Updated

File: datasync/downloader/downloader.go [hunk 9]
```diff
@@ -447,10 +450,17 @@ func (d *Downloader) syncWithPeer(p *peerConnection, hash common.Hash, td *big.I
...
+	latest, pivot, err := d.fetchHead(p)
 	if err != nil {
 		return err
 	}
+	if mode == FastSync && pivot == nil {
+		// If no pivot block was returned, the head is below the min full block
+		// threshold (i.e. new chain). In that case we won't really fast sync
+		// anyway, but still need a valid pivot block to avoid some code hitting
+		// nil panics on an access.
+		pivot = d.blockchain.CurrentBlock().Header()
```
[the remark this answers is not in the document]
Comment (jeongkyun-oh) [reply datasync/downloader/downloader.go:462]: When `fetchHead` returns the pivot as nil, it sets `pivot = d.blockchain.CurrentBlock().Header()`. Do I understand your point correctly?

**Review by kjhman21 — APPROVED** [review]

**Review by aidan-kwon — APPROVED** [review]

**Review by ethan-kr — APPROVED** [review]
